Add focus check to onTapUpOutside

Hannnes1 opened the pull request:
Adds a focus check to `onTapUpOutside` of `EditableText`, so that it doesn't trigger if the `EditableText` doesn't have focus. 

`onTapOutside` already had this check, but it was missed when `onTapUpOutside` was added.

Fixes https://github.com/flutter/flutter/issues/162573


## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

Hannnes1 committed:
Add focus check to onTapUpOutside

Hannnes1 commented on packages/flutter/test/widgets/editable_text_test.dart:
```diff
@@ -16877,6 +16877,80 @@ void main() {
+  });
+
+  // Regression test for https://github.com/flutter/flutter/issues/162573
+  testWidgets('onTapUpOutside is not called upon tap up outside when field is not focused', (
```
The equivalent test for `onTapOutside` is located in [text_form_field_test.dart#L741](https://github.com/Mairramer/flutter/blob/b4a81cb6da84b7658b984922fb7056920058a14e/packages/flutter/test/material/text_form_field_test.dart#L741), but in my opinion it should have been located in this file instead. That's why I added this test here. Let me know if I should move it to `text_form_field_test.dart` instead.

Renzo-Olivares replied:
I think this test is good in `editable_text_test.dart`. If you'd like you can also transfer the equivalent `onTapOutside` test from `text_form_field_test.dart` to `editable_text_test.dart`, I think it fits better in here.

Renzo-Olivares reviewed (approved):
LGTM, thank you for the contribution!

Hannnes1 committed:
Prevent tap down from blocking tap up in test

hannah-hyj reviewed (approved):
LGTM!

Renzo-Olivares commented on packages/flutter/test/widgets/editable_text_test.dart:
```diff
@@ -11987,13 +11994,18 @@ void main() {
             Actions(
-              actions: <Type, Action<Intent>>{EditableTextTapUpOutsideIntent: overrideAction},
+              actions: <Type, Action<Intent>>{
+                // The tap down has to be overridden so that the tap up will register.
```
Is this the expected behavior? I would expect that I can override `EditableTextTapUpOutsideIntent` without overriding `EditableTextTapOutsideIntent`.

Hannnes1 replied:
The default behavior of `EditableTextTapOutsideIntent` is to unfocus the text field (except on mobile touch events). That means that the text field would be unfocused on the tap up event, blocking `EditableTextTapUpOutsideIntent`.
But I see how that might not be intuitive. I see a few solutions:
* Document the current behavior better
* Remove the focus check from `onTapUpOutside` (Close this PR)
* Check if the text field had focus previously

I'm open to other solutions as well

Renzo-Olivares replied:
> Remove the focus check from onTapUpOutside (Close this PR)
If we keep the focus check, then `onTapOutside` will never fire unless `onTapOutside` is overridden. If we remove it, this means every `TextField` that doesn't have focus and was "tapped outside" will receive a "onTapUpOutside", this doesn't seem like the correct behavior. I think it makes sense that only the `TextField` that was focused/previously focused receives "tapped outside" events.

> Check if the text field had focus previously
This sounds reasonable to me. The focus should probably be checked in `onTapOutside`, before we unfocus.

Hannnes1 committed:
Trigger onTapUpOutside based on focus state on tap down

Hannnes1 committed:
Move and add test

Moved one test that should have been in editable_text_test.
Also added one test to make sure I didn't break onTapOutside.

Hannnes1 committed:
Format

Hannnes1 commented:
@Renzo-Olivares I added a `_hadFocusOnTapDown` that fixes the issue with `onTapOutside` blocking `onTapUpOutside`

justinmc commented:
@Hannnes1 FYI there are some test failures now.

Hannnes1 commented:
Thanks.
Haven't figured out why yet, but will look into it

Hannnes1 committed:
Fix a failing test on web

justinmc commented:
Looks like the semantics tree doesn't match any more. I see the flags `SemanticsFlag.hasRequiredState` and
`SemanticsFlag.isRequired` are missing now I think? Not sure if that's intended or not.

Renzo-Olivares commented:
I know there have been a lot of semantic changes recently, maybe rebasing will fix some of these?

Hannnes1 commented:
> I know there have been a lot of semantic changes recently, maybe rebasing to master will fix some of these?

Rebasing again worked. Thanks!
